supervisor/shared/display.c

```diff
@@ -78,8 +78,20 @@ void supervisor_start_terminal(uint16_t width_px, uint16_t height_px) {
 
     uint16_t total_tiles = width_in_tiles * height_in_tiles;
 
-    // First try to allocate outside the heap. This will fail when the VM is running.
-    tilegrid_tiles = allocate_memory(align32_size(total_tiles), false);
+    // Reuse the previous allocation if possible
+    if (tilegrid_tiles) {
```

Comment:
The commit message tried to explain it:

> If `supervisor_start_terminal` is called twice in a row without
> `supervisor_stop_terminal`, it would lose track of a supervisor
> allocation.
> 
> This can occur when setting the rotation of a display, including the
> way that a FramebufferDisplay sets rotation _AFTER_ initial construction,
> first with a stack like
> ```
> #0  supervisor_srtart_terminal
> #1  in displayio_display_core_construct
> #2  in common_hal_framebufferio_framebufferdisplay_construct
> ```
> and then with a stack like
> ```
> #0  supervisor_start_terminal
> #1  in common_hal_framebufferio_framebufferdisplay_construct
> #2  in framebufferio_framebufferdisplay_make_new
> ```
> .. without an intervening stop_terminal call.
> 
> For reasons I didn't fully explore, this did not become a problem until
> the ability to re-allocate a freed supervisor allocation was
> implemented in #3482.  Demonstrating the problem requires #3482 + #3497
> + this PR.

However, now I'm questioning this line of reasoning and I wonder if common_hal_displayio_display_construct and common_hal_framebufferio_framebufferdisplay_construct are simply wrong to call `supervisor_start_terminal` directly.  This call was added in 1a1dbef992442682413f6103e36440a26bc2dad7 (displayio) and blindly carried over to framebufferio.  As far as I can see It's been there forever, but maybe should have been removed when the call was added in displayio_display_core_construct during the refactor that created it.  I can test this and change the solution if it works.

Reply:
:+1: Seems to me you'll want to remove the direct call in favor of the one done by core_construct.

Reply:
This was fixed in a better way in another PR, so it's been rebased out of here